### PR TITLE
Check for valid dependency

### PR DESCRIPTION
**Pull request by vinit717**
**Overview**

- Currently, we can pass any string as dependency which may cause bad behavior So for valid dependency(taskid) we have to check that it exists or not  

**Solution**

- We can check as dependency by  checking doc id of task model exist or not if exist we can take as dependency otherwise throw an error

image

### CODE REVIEW

File: test/integration/tasks.test.js [hunk 1]
```diff
@@ -328,6 +329,18 @@ describe("Tasks", function () {
 
       return taskId;
     });
+
+    it("Should return 400 if taskid is not exist", async function () {
+      taskId = (await tasks.updateTask(tasksData[5])).taskId;
```
Comment (heyrandhir) [inline test/integration/tasks.test.js:334]: what if 1 task is valid and 1 in valid ?

File: models/tasks.js [hunk 2]
```diff
@@ -33,15 +33,19 @@ const updateTask = async (taskData, taskId = null) => {
...
                 dependsOn: dependency,
               };
-              const docRef = dependencyModel.doc();
-              transaction.set(docRef, taskDependsOn);
+              if (dependencyDoc.exists) {
+                const docRef = dependencyModel.doc();
+                transaction.set(docRef, taskDependsOn);
+              } else {
+                throw new Error("Invalid dependency taskId");
+              }
```
Comment (heyrandhir) [inline models/tasks.js:48]: I don't see model tests getting updated. Please check

**Review by heyrandhir — CHANGES_REQUESTED** [review]

File: test/unit/models/tasks.test.js [hunk 3]
```diff
@@ -6,15 +6,15 @@
 
 const chai = require("chai");
 const { expect } = chai;
-
+// const sinon = require("sinon");
```
Comment (heyrandhir) [inline test/unit/models/tasks.test.js:9]: please remove this if not needed

File: test/integration/tasks.test.js [hunk 4]
```diff
@@ -311,15 +310,17 @@ describe("Tasks", function () {
     });
     it("Should update dependency", async function () {
       taskId = (await tasks.updateTask(tasksData[5])).taskId;
-      const dependsOn = ["taskId5", "taskId4"];
+      const taskId5 = (await tasks.updateTask(tasksData[5])).taskId;
+      const taskId6 = (await tasks.updateTask(tasksData[5])).taskId;
```
Comment (heyrandhir) [inline test/integration/tasks.test.js:314]: why 5 and 6  in place of 1 and 2 🤔

File: test/integration/tasks.test.js [hunk 5]
```diff
@@ -328,6 +329,18 @@ describe("Tasks", function () {
...
+    it("Should return 400 if taskid is not exist", async function () {
+      taskId = (await tasks.updateTask(tasksData[5])).taskId;
+
+      const dependsOn = ["taskId5", "taskId6"];
+      const res = await chai
+        .request(app)
+        .patch(`/tasks/${taskId}`)
+        .set("cookie", `${cookieName}=${jwt}`)
+        .send({ dependsOn });
+      expect(res).to.have.status(400);
```
Comment (heyrandhir) [inline test/integration/tasks.test.js:342]: can we also check the error message

File: models/tasks.js (hunk 2 above)
Comment (vinit717) [reply models/tasks.js]: Added

File: test/unit/models/tasks.test.js [hunk 6]
```diff
@@ -140,4 +145,31 @@ describe("tasks", function () {
...
+      await dependencyModel.doc("taskDependencies").set(data);
+      const dependencyData = (await dependencyModel.doc("taskDependencies").get()).data();
+
+      expect(dependencyData.dependsOn).to.be.a("array");
+      expect(dependencyData.taskId).to.be.equal("taskId1");
+
+      setStub.restore();
+      getStub.restore();
+    });
+  });
```
Comment (heyrandhir) [inline test/unit/models/tasks.test.js:174]: where is the model function being called here

**Review by heyrandhir — CHANGES_REQUESTED** [review]

File: test/unit/models/tasks.test.js (hunk 6 above)
Comment (vinit717) [reply test/unit/models/tasks.test.js]: Added

File: test/integration/tasks.test.js (hunk 5 above)
Comment (vinit717) [reply test/integration/tasks.test.js:342]: Added

File: test/integration/tasks.test.js (hunk 4 above)
Comment (vinit717) [reply test/integration/tasks.test.js]: Resolved

File: test/unit/models/tasks.test.js (hunk 3 above)
Comment (vinit717) [reply test/unit/models/tasks.test.js]: Removed

File: test/integration/tasks.test.js (hunk 1 above)
Comment (vinit717) [reply test/integration/tasks.test.js]: Done

**Review by heyrandhir — CHANGES_REQUESTED** [review]

File: test/unit/models/tasks.test.js [hunk 7]
```diff
@@ -140,4 +144,31 @@ describe("tasks", function () {
...
+      const data = {
+        taskId: "taskId1",
+        dependsOn: ["taskId2", "taskId3"],
+      };
+
+      const setStub = sinon.stub(dependencyModel.doc("taskDependencies"), "set").resolves();
+      const getStub = sinon.stub(dependencyModel.doc("taskDependencies"), "get").resolves({
+        data: () => data,
+      });
+      const result = await updateTask(data);
```
[the remark this answers is not in the document]
Comment (vinit717) [reply test/unit/models/tasks.test.js:163]: Yes when we pass the update task it returns taskid this model function also used to create a task which return task object

File: controllers/tasks.js [hunk 8]
```diff
@@ -205,8 +205,14 @@ const updateTask = async (req, res) => {
     if (isUserStatusEnabled && req.body.assignee) {
       await updateUserStatusOnTaskUpdate(req.body.assignee);
     }
+
     return res.status(204).send();
   } catch (err) {
+    if (err.message.includes("Invalid dependency taskId")) {
```
[the remark this answers is not in the document]
Comment (vinit717) [reply controllers/tasks.js]: done

File: test/unit/models/tasks.test.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (heyrandhir) [reply test/unit/models/tasks.test.js]: `taskId` is string. Right ?

[the remark this answers is not in the document]
Comment (vinit717) [reply test/unit/models/tasks.test.js]: yes also can you re-review the test I had changed the tests

**Review by heyrandhir — APPROVED** [review]

**Review by kotesh-arya — APPROVED** [review]

**Pull request merged**